modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionDemander.java

```diff
@@ -1647,7 +1647,8 @@ private void sendRebalanceFinishedEvent() {
          * @return {@code True} when future compared with other, {@code False} otherwise.
          */
         public boolean compatibleWith(GridDhtPreloaderAssignments otherAssignments) {
-            if (isInitial() || !allNodesSupports(ctx.kernalContext(), otherAssignments.keySet(), TX_TRACKING_UPDATE_COUNTER))
+            if (isInitial() || !allNodesSupports(ctx.kernalContext(), otherAssignments.keySet(), TX_TRACKING_UPDATE_COUNTER)
+                || ((GridDhtPreloader)grp.preloader()).isDisableRebalancingCancellationOptimization())
```

Comment:
Better place it first in conditions order.

Reply:
Haven't possible to put it first, because init future does not have context (all fields are null).